x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java

```diff
@@ -51,54 +62,142 @@ int getPositionCount() {
-                    case 0 -> null;
-                    case 1 -> field.termQuery(blockValueReader.apply(first), searchExecutionContext);
-                    default -> {
-                        final List<Object> terms = new ArrayList<>(count);
-                        for (int i = 0; i < count; i++) {
-                            final Object value = blockValueReader.apply(first + i);
-                            terms.add(value);
-                        }
-                        yield field.termsQuery(terms, searchExecutionContext);
+        return new TermQueryList(field, searchExecutionContext, block);
+    }
+
+    /**
+     * Returns a list of geo_shape queries for the given field and the input block.
+     */
+    static QueryList geoShapeQuery(MappedFieldType field, SearchExecutionContext searchExecutionContext, Block block) {
+        return new GeoShapeQueryList(field, searchExecutionContext, block);
+    }
+
+    private static class TermQueryList extends QueryList {
+        private final BytesRef scratch = new BytesRef();
+        private final byte[] ipBytes = new byte[InetAddressPoint.BYTES];
+        private final MappedFieldType field;
+        private final SearchExecutionContext searchExecutionContext;
+        private final IntFunction<Object> blockValueReader;
+
+        private TermQueryList(MappedFieldType field, SearchExecutionContext searchExecutionContext, Block block) {
+            super(block);
+
+            this.field = field;
+            this.searchExecutionContext = searchExecutionContext;
+            this.blockValueReader = blockToJavaObject();
+        }
+
+        @Override
+        Query getQuery(int position) {
+            final int first = block.getFirstValueIndex(position);
+            final int count = block.getValueCount(position);
+            return switch (count) {
+                case 0 -> null;
+                case 1 -> field.termQuery(blockValueReader.apply(first), searchExecutionContext);
+                default -> {
+                    final List<Object> terms = new ArrayList<>(count);
+                    for (int i = 0; i < count; i++) {
+                        final Object value = blockValueReader.apply(first + i);
+                        terms.add(value);
                     }
-                };
-            }
-        };
+                    yield field.termsQuery(terms, searchExecutionContext);
+                }
+            };
+        }
+
+        private IntFunction<Object> blockToJavaObject() {
+            return switch (block.elementType()) {
+                case BOOLEAN -> {
+                    BooleanBlock booleanBlock = (BooleanBlock) block;
+                    yield booleanBlock::getBoolean;
+                }
+                case BYTES_REF -> {
+                    BytesRefBlock bytesRefBlock = (BytesRefBlock) block;
+                    if (field instanceof RangeFieldMapper.RangeFieldType rangeFieldType && rangeFieldType.rangeType() == RangeType.IP) {
```

Comment:
I think we should support strings and ip values (IP ranges later). This means we should handle the conversion based on the data type of the input block, not the data type of the match field in the enrich index. To implement this, we'll need another prerequisite PR to include the data type of the input block in LookupRequest within EnrichLookupService.

Reply:
I assume also we should do a validation on the data types of both the input block and the match field? Right now in Analyzer we only check that the input data type is compatible with the match type (eg. `geo_match` supports only `geo_point` and `geo_shape`), but we don't check the input type against the match field type, which we could if we passed in the match field type like you suggest.

Reply:
I've added the input dataType support in a separate PR at https://github.com/elastic/elasticsearch/pull/106344

Reply:
I've pushed a commit that now uses the input data type to decide on how to read the input block, which I agree makes more sense. I'm only using this for DataTypes.IP. You suggested supporting strings also. Does that mean you want to bring back support for enriching strings against IP ranges? Right now if the source index contains strings, not IP's the user would need to explicitly `EVAL ip=to_ip(str)` before doing the ENRICH. This seems like a good idea, rather that directly supporting strings inside the ENRICH.

Reply:
As a test, I reduced the strictness on non-geo types for match/range.

This also allows keywords to be used in ip_range matches, which was the previous behaviour. However, this also means that invalid IP strings will throw errors. I added a test asserting on that error, and think this matches previous behaviour, but wonder if we should instead return null and set a warning.